src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1084 @@
+        // However, for observability reasons, it would be much more readble if the shard ID was
+        // constructed as `format!("s{}CATALOG", hash(environment_id))`.
+        let shard_id = ShardId::from_str(&format!("s{environment_id}")).expect("known to be valid");
+        let (write_handle, read_handle) = persist_client
+            .open(
+                shard_id,
+                Arc::new(VecU8Schema::default()),
+                Arc::new(UnitSchema),
+                Diagnostics {
+                    shard_name: "catalog".to_string(),
+                    handle_purpose: "durable catalog state".to_string(),
+                },
+            )
+            .await
+            .expect("invalid usage");
+        PersistHandle {
+            write_handle,
+            read_handle,
+        }
+    }
+
+    async fn open_inner(
+        mut self,
+        mode: Mode,
+        now: NowFn,
+        bootstrap_args: &BootstrapArgs,
+        deploy_generation: Option<u64>,
+    ) -> Result<PersistCatalogState, CatalogError> {
+        let (is_initialized, upper) = self.is_initialized_inner().await;
+
+        if !matches!(mode, Mode::Writable) && !is_initialized {
+            return Err(CatalogError::Durable(DurableCatalogError::NotWritable(
+                format!("catalog tables do not exist; will not create in {mode:?} mode"),
+            )));
+        }
+
+        let version = self.get_user_version(upper).await;
+        let read_only = matches!(mode, Mode::Readonly);
+
+        let mut initial_snapshot: Vec<_> = self.snapshot(upper).await.collect();
+        let mut epoch = if is_initialized {
+            // Sniff out the most recent epoch.
+            let epoch_idx = initial_snapshot
+                .iter()
+                .rposition(|update| {
+                    matches!(
+                        update,
+                        StateUpdate {
+                            kind: StateUpdateKind::Epoch(_),
+                            ..
+                        }
+                    )
+                })
+                .expect("initialized catalog must have an epoch");
+            match initial_snapshot.remove(epoch_idx) {
+                StateUpdate {
+                    kind: StateUpdateKind::Epoch(epoch),
+                    diff,
+                    ..
+                } => {
+                    soft_assert_eq!(diff, 1);
+                    epoch
+                }
+                _ => unreachable!("checked above"),
+            }
+        } else {
+            Epoch::new(1).expect("known to be non-zero")
+        };
+        // Note only writable catalogs attempt to increment the epoch.
+        if matches!(mode, Mode::Writable) {
+            epoch = Epoch::new(epoch.get() + 1).expect("known to be non-zero");
+        }
+
+        let mut catalog = PersistCatalogState {
+            mode,
+            persist_handle: self,
+            upper,
+            epoch,
+            snapshot: Snapshot::empty(),
```

Comment:
Just making sure I understand, once the Persist backed version of the Stash is using our protobufs, will we still use an empty snapshot here?

Reply:
I've added some extra comments to clarify.